compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4958,7 +4958,7 @@ public void visit(BLangTypeConversionExpr conversionExpr) {
             return;
         }
         conversionExpr.typeNode = rewrite(conversionExpr.typeNode, env);
-        if (conversionExpr.type.tag == TypeTags.STRING && conversionExpr.expr.type.tag == TypeTags.XML_TEXT) {
+        if (conversionExpr.type.tag == TypeTags.STRING && types.isXMLConversionExprCastableToString(conversionExpr)) {
```

Comment:
I think we need to improve this a bit to handle scenarios where the target may also be a union. Basically we need a method that accepts the source and the target, and checks if the target has a subtype of `string` and the source has a subtype of `xml:Text`.

Reply:
The following also doesn't work atm.

```ballerina
type FooBar "foo"|"bar";

public function main() {
    'xml:Text t = xml `foo`;
    FooBar u = <FooBar> t;
}
```